src/main/java/frc/robot/subsystems/Swerve/SwerveSubsystem.java

```diff
@@ -0,0 +1,156 @@
+        gyro.configFactoryDefault();
+        zeroGyro();
+
+        mSwerveMods = new SwerveModule[] {
+            new SwerveModule(0, Constants.Swerve.Mod0.constants),
+            new SwerveModule(1, Constants.Swerve.Mod1.constants),
+            new SwerveModule(2, Constants.Swerve.Mod2.constants),
+            new SwerveModule(3, Constants.Swerve.Mod3.constants)
+        };
+        //TODO: Set the actual pose
+        swerveOdometry = new SwerveDrivePoseEstimator(Constants.Swerve.swerveKinematics, getYaw(), getModulePositions(), new Pose2d());
+    }
+
+    public void drive(Translation2d translation, double rotation, boolean fieldRelative, boolean isOpenLoop) {
+        SwerveModuleState[] swerveModuleStates =
+            Constants.Swerve.swerveKinematics.toSwerveModuleStates(
+                fieldRelative ? ChassisSpeeds.fromFieldRelativeSpeeds(
+                                    translation.getX(), 
+                                    translation.getY(), 
+                                    rotation, 
+                                    getYaw()
+                                )
+                                : new ChassisSpeeds(
+                                    translation.getX(), 
+                                    translation.getY(), 
+                                    rotation)
+                                );
+        SwerveDriveKinematics.desaturateWheelSpeeds(swerveModuleStates, Constants.Swerve.maxSpeed);
+
+        for(SwerveModule mod : mSwerveMods){
+            mod.setDesiredState(swerveModuleStates[mod.moduleNumber], isOpenLoop);
+        }
+    }    
+
+        /**
+     * Moves the drivebase around by running the swerve modules.
+     * 
+     * @param chassisSpeeds The x, y, and theta the drivebase must move in.
+     */
+    public void drive(ChassisSpeeds chassisSpeeds) {
+        this.chassisSpeeds = chassisSpeeds;
+    }
+
+    public void drive(Trajectory.State targetState, Rotation2d targetRotation) {
+        // determine ChassisSpeeds from path state and positional feedback control from
+        // HolonomicDriveController
+        lastGivenRotation = targetRotation;
+        ChassisSpeeds targetChassisSpeeds = pathController.calculate(
+                getPose(),
+                targetState,
+                targetRotation);
+        // command robot to reach the target ChassisSpeeds
+        drive(targetChassisSpeeds);
+    }
+
+    /* Used by SwerveControllerCommand in Auto */
+    public void setModuleStates(SwerveModuleState[] desiredStates) {
+        SwerveDriveKinematics.desaturateWheelSpeeds(desiredStates, Constants.Swerve.maxSpeed);
+        
+        for(SwerveModule mod : mSwerveMods){
+            mod.setDesiredState(desiredStates[mod.moduleNumber], false);
+        }
+    }    
+
+    public Pose2d getPose() {
+        return swerveOdometry.getEstimatedPosition();
+    }
+
+    public void addVisionMeasurement(Pose2d pose, double latency) {
```

Comment:
This can also be private, this should only be called internally.

Reply:
Fixed!